### ISSUE TITLE
When jumping from native iOS to FlutterViewController sometimes the UI freezes.

### ISSUE THREAD

**Issue by 386212519** [flutter/flutter#39036]
I used a communication plugin, ```hybrid_stack_manager```, to ensure communication between native and flutter. When jumping from native iOS to FlutterViewController, sometimes the UI freezes. The tracer found that the code went to the ```ViewWillAppear``` method of ```flutterViewController``` but did not take the ```viewDidAppear``` method. Viewing the layer found that the ```flutterViewController``` layer is already on the window, but the previous layer has not fully launched the ```NavigationController```. The page now displays the previous native page UI. When I press the home button to push the app to the background, then enter the app again and the UI for the flutter page is displayed. I have been looking for a long time and I am not sure what the problem is. Has anyone encountered such a problem? Can you share your solution? thank you very much!

## Flutter Doctor
```
[✓] Flutter (Channel stable, v1.7.8, on Mac OS X 10.14.5 18F132, locale zh-Hans-CN)
[!] Android toolchain - develop for Android devices (Android SDK version 28.0.3)
    ✗ Android license status unknown.
      Try re-installing or updating your Android SDK Manager.
      See https://developer.android.com/studio/#downloads or visit https://flutter.dev/setup/#android-setup for detailed instructions.
[✓] Xcode - develop for iOS and macOS (Xcode 10.2.1)
[✓] iOS tools - develop for iOS devices
[!] Android Studio (version 3.2)
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
[✓] VS Code (version 1.37.1)
[✓] Connected device (1 available)

! Doctor found issues in 2 categories.
```

**Comment by BondarenkoStas** [comment]
@386212519 Please provide a code snippet to be sure to reproduce it correctly.

**Comment by jaysephjw** [comment]
I was about to create my own ticket before seeing this one. My issue is: "Drawing freezes when switching between two FlutterViewControllers in an iOS UITabBar". 

I can reproduce 100% on simulator and iPhone 7.

Here's a minimal repro repo! https://github.com/jaysephjw/tab_switch_example

I'm not using any plugins, and reproduced without platform channels. 

I setup an iOS UITabBar, which is a bottom bar that switches between View Controllers, like so:
```
        platformVC = UIViewController(nibName: nil, bundle: nil)
        flutterVC1 = FlutterViewController(engine: flutterEngine, nibName: nil, bundle: nil)
        flutterVC2 = FlutterViewController(engine: flutterEngine, nibName: nil, bundle: nil)
       ... 
        self.viewControllers = [platformVC, flutterVC1, flutterVC2] 
```

Then, to control attachment of the engine, I use this UITabControllerDelegate method:
```
    // Watch for when the user switches to a FlutterVC. When they do, attach it to the engine.
    override func tabBar(_ tabBar: UITabBar, didSelect item: UITabBarItem) {
        if (item.title == "Flutter 1") {
            flutterEngine!.viewController = flutterVC1
        } else if (item.title == "Flutter 2") {
            flutterEngine!.viewController = flutterVC2
        }
    }
```

Here's what happens: 
![flutter_tabs_freeze](https://user-images.githubusercontent.com/3268245/64365559-471d7600-cfc9-11e9-98b9-c41ebcd9f2ac.gif)


Does my code in `didSelect` miss something? Are steps needed to move the engine around besides setting its viewController property?

My goal is to use Flutter in two different platform tabs. Any help appreciated! Hope this helps diagnose too.

**Comment by jaysephjw** [comment]
Found a work around for my issue.

Wrap each FlutterViewController in another VC. Add/Remove the flutter VCs when the tab is shown, like so: 
```
    override func viewDidAppear(_ animated: Bool) {
        embedChild() // add flutterview as child VC
        
       if (child.engine.viewController != child) {
            child.engine.viewController = child
        }
    }
    
    override func viewWillDisappear(_ animated: Bool) {
        removeChild()
    }
```

Haven't figured out why this works. **One insight is that setting the engine's VC to the current VC causes a freeze.** Could be related. Guarding against this is necessary but not sufficient to avoiding the freeze switching tabs; for some reason detatching then reattching the flutter VCs from the view hierarchy is necessary. 

Working example at https://github.com/jaysephjw/tab_switch_example/blob/wrapped_fix/ios/Runner/CustomUITabBar.swift

btw @386212519 consider trying out https://github.com/alibaba/flutter_boost instead of hybrid_router, it looks more up to date.

**Comment by BondarenkoStas** [comment]
@jaysephjw Thank you a lot for investigation! 
@386212519 Please report it that helps.

**Comment by 386212519** [comment]
> 为我的问题找到了解决方法。
> 
> 将每个FlutterViewController包装在另一个VC中。显示选项卡时添加/删除颤动VC，如下所示：
> 
> ```
> class WrapperVC : UIViewController {
> ...
>     override func viewDidAppear(_ animated: Bool) {
>         embedChild() // add flutterview as child VC
>         
>        if (child.engine.viewController != child) {
>             child.engine.viewController = child
>         }
>     }
>     
>     override func viewWillDisappear(_ animated: Bool) {
>         removeChild()
>     }
> }
> ```
> 
> 还没弄清楚为什么会这样。**一个见解是将引擎的VC设置为当前VC会导致冻结。**可能有关系。防止这种情况是必要的，但不足以避免冻结切换标签; 由于某种原因，需要进行分离，然后从视图层次中重新选择颤振VC。
> 
> [https://github.com/jaysephjw/tab_switch_example/blob/wrapped_fix/ios/Runner/CustomUITabBar.swift上的](https://github.com/jaysephjw/tab_switch_example/blob/wrapped_fix/ios/Runner/CustomUITabBar.swift)工作示例[](https://github.com/jaysephjw/tab_switch_example/blob/wrapped_fix/ios/Runner/CustomUITabBar.swift)
> 
> btw @ 386212519考虑尝试https://github.com/alibaba/flutter_boost而不是hybrid_router，它看起来更新了。
Thank you very much, I am using the third-party plugin ‘hybrid_stack_manager’ which is what you said. ‘Wrap each FlutterViewController in another VC’ to render FLutterViewController. But there is still this problem. I have consulted several friends who have developed with FLutter, and they have suggested that I use ‘flutter_boost’ to develop. Only if I change, both Android and web will change. So still considering it.

**Comment by 386212519** [comment]
> @jaysephjw非常感谢您的调查！
> @ 386212519请报告有帮助。

Thank you very much, the third-party plug-in we are currently using is developed by the way upstairs friends. Just still have problems. I found it when I tested it many times. It's very likely that when I swipe the finger on the screen several times, then jumping to the FlutterviewController will have a high chance of freezing the UI. However, I still need to go through multiple tests to determine this situation. I am currently suspected to be caused by a third-party plugin. But it can't be ruled out that flutter has a problem when rendering the UI.

**Comment by jmagman** [comment]
There was a lot of performance work done on this area that landed in v1.9.1.  Do you still see the UI freeze on the latest stable version of Flutter?

**Comment by iapicca** [comment]
Without additional information we are unfortunately not sure how to resolve this issue. 
We are therefore reluctantly going to close this bug for now. 
Please don't hesitate to comment on the bug if you have any more information for us; we will reopen it right away!  
Thanks for your contribution.

Could everyone who still has this problem please file a new issue with the exact description of what happens, logs and the output of `flutter doctor -v`.
All system setups can be slightly different, so it's always better to open new issues and reference related issues.
